design-proposals/instancetype.kubevirt.io/extend-preferSpread.md

```diff
@@ -0,0 +1,325 @@
+metadata:
+  name: spread
+spec:
+  cpu:
+    preferredCPUTopology: Spread
+---
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  name: vm-spread
+spec:
+  instancetype:
+    name: spread
+    kind: VirtualMachineInstancetype
+  preference:
+    name: spread
+    kind: VirtualMachinePreference
+  runStrategy: Always
+  template:
+    spec:
+      domain:
+        devices: {}
+      volumes:
+        - containerDisk:
+            image: quay.io/containerdisks/fedora:39
+          name: containerdisk
+EOF
+[..]
+$ ./cluster-up/kubectl.sh get vmi/vm-spread -o json | jq .spec.domain.cpu
+{
+  "cores": 4,
+  "sockets": 2,
+  "threads": 1
+}
+```
+
+### `CoresThreads`
+
+```yaml
+$ ./cluster-up/kubectl.sh apply -f -<<EOF
+---
+apiVersion: instancetype.kubevirt.io/v1beta1
+kind: VirtualMachineInstancetype
+metadata:
+  name: spread
+spec:
+  cpu:
+    guest: 8
+  memory:
+    guest: 1Gi
+---
+apiVersion: instancetype.kubevirt.io/v1beta1
+kind: VirtualMachinePreference
+metadata:
+  name: spread
+spec:
+  cpu:
+    preferredCPUTopology: Spread
+    spreadOptions:
+      across: CoresThreads
+---
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  name: vm-spread
+spec:
+  instancetype:
+    name: spread
+    kind: VirtualMachineInstancetype
+  preference:
+    name: spread
+    kind: VirtualMachinePreference
+  runStrategy: Always
+  template:
+    spec:
+      domain:
+        devices: {}
+      volumes:
+        - containerDisk:
+            image: quay.io/containerdisks/fedora:39
+          name: containerdisk
+EOF
+[..]
+$ ./cluster-up/kubectl.sh get vmi/vm-spread -o json | jq .spec.domain.cpu
+{
+  "cores": 4,
+  "sockets": 1,
+  "threads": 2
+}
+```
+
+### `SocketsCoresThreads`
+
+```yaml
+$ ./cluster-up/kubectl.sh apply -f -<<EOF
+---
+apiVersion: instancetype.kubevirt.io/v1beta1
+kind: VirtualMachineInstancetype
+metadata:
+  name: spread
+spec:
+  cpu:
+    guest: 16
+  memory:
+    guest: 1Gi
+---
+apiVersion: instancetype.kubevirt.io/v1beta1
+kind: VirtualMachinePreference
+metadata:
+  name: spread
+spec:
+  cpu:
+    preferredCPUTopology: Spread
+    spreadOptions:
+      across: SocketsCoresThreads
+---
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  name: vm-spread
+spec:
+  instancetype:
+    name: spread
+    kind: VirtualMachineInstancetype
+  preference:
+    name: spread
+    kind: VirtualMachinePreference
+  runStrategy: Always
+  template:
+    spec:
+      domain:
+        devices: {}
+      volumes:
+        - containerDisk:
+            image: quay.io/containerdisks/fedora:39
+          name: containerdisk
+EOF
+[..]
+$ ./cluster-up/kubectl.sh get vmi/vm-spread -o json | jq .spec.domain.cpu
+{
+  "cores": 4,
+  "sockets": 2,
+  "threads": 2
+}
+```
+
+## Functional Testing Approach
+
+* Extend existing unit test coverage to ensure the expected topologies are generated within the VirtualMachineInstance.
+* Unclear if any additional functional test coverage is required outside of this.
```

Comment:
Tests / verification that deny SocketsThreads without dedicated CPU placement?

Reply:
See https://github.com/kubevirt/kubevirt/issues/11749 - I wouldn't enforce that here.